### PR TITLE
Add Groestlcoin (GRS) support

### LINKED ISSUE #273: Add support for Groestlcoin (GRS) addresses in encoding library
**Previous PRs to add GRS have been [rejected due to increasing the file size of the library too much](https://github.com/ensdomains/address-encoder/pull/239). We need someone who can figure out how to add GRS with as little space as possible

Here is the encoding library: https://github.com/ensdomains/address-encoder

Example: https://github.com/ensdomains/address-encoder/pull/47/files

Discord for dev help: https://discord.com/invite/AskZbFx

Please provide a reference link so we can check you did the encoding correctly.

We'd like to keep the encoding library as small as possible, so please be efficient in your use of space.

We will likely be reviewing PRs and paying out bounties more or less on a weekly basis. No need to ping us to check your PR, we will definitely get to it.

Thanks for your help!

### PR DESCRIPTION
**Pull request by janus**
This PR provides encoding and decoding support for Groestlcoin addresses
## Issue number
closes #273

## Description

Adds groestl-hash-js lib to package.json, and creates grsCheckSumFn, bs58grscheckEncode, decodeRaw, bs58grscheckDecode, makeBase58GrsCheckEncoder, makeBase58GrsCheckDecoder, makeGroestlcoinEncoder, makeGroestlcoinDecoder and  groestlcoinChain functions in index.ts file. Adds test cases related to GRS inside index.test.ts

## Reference to the specification

- https://github.com/trustwallet/wallet-core/blob/master/src/Groestlcoin/Address.cpp
- https://github.com/Groestlcoin/groestlcoin/blob/master/src/wallet/wallet.cpp

## Reference to the test address.
 - Test done according to test setup
 - https://chainz.cryptoid.info/grs/

## List of features added/changed

- base58check P2PKH and P2SH, and bech32 segwit for GRS

NOTE: If you are adding new coin address support, please make sure to add a reference link to README so that reviewer can verify.

## How much has the filesize increased?

457592

## How Has This Been Tested?
Tested and passed all checks.

## Checklist:

- [x] My code follows the code style of this project.
- [x] My code implements all the required features.
- [x] I have specified correct coinTypes specified at [Slip 44](https://github.com/satoshilabs/slips/blob/master/slip-0044.md)
- [x] I have provided the reference link to the specification I implemented.
- [x] I have provided enough explanation about how I provided the test address

### CODE REVIEW

**Comment by makoto** [comment]
345159 -> 457592 is over 100k increase. I think it's too big

**Comment by janus** [comment]
> 345159 -> 457592 is over 100k increase. I think it's too big

Current size: 376465

**Comment by janus** [comment]
@makoto  Resolved. Review.

**Comment by makoto** [comment]
Your linting is still failing

File: src/index.ts [hunk 1]
```diff
@@ -248,6 +249,123 @@ function decodeBitcoinCash(data: string): Buffer {
...
+  p2pkhVersions: base58CheckVersion[],
+  p2shVersions: base58CheckVersion[],
+) => ({
+  coinType,
+  decoder: makeGroestlcoinDecoder(hrp, p2pkhVersions, p2shVersions),
+  encoder: makeGroestlcoinEncoder(hrp, p2pkhVersions[0], p2shVersions[0]),
+  name,
+});
+
+
```
Comment (makoto) [inline src/index.ts:366]: I noticed that you tend to leave lots of blank lines. Can you tidy up?

File: src/groestl-hash-js/groestl.js [hunk 2]
```diff
@@ -0,0 +1,202 @@
+/////////////////////////////////////
```
Comment (makoto) [inline src/groestl-hash-js/groestl.js:1]: Can you add a comment mentioning where you copied the files from?

**Comment by makoto** [comment]
Now that you ported depending library to own files, do you still need to commit changes on `package-lock.json` ?

File: src/index.ts (hunk 1 above)
Comment (janus) [reply src/index.ts]: Done.

File: src/groestl-hash-js/groestl.js (hunk 2 above)
Comment (janus) [reply src/groestl-hash-js/groestl.js:1]: Added.

**Comment by janus** [comment]
> Now that you ported depending library to own files, do you still need to commit changes on `package-lock.json` ?

Thanks for pointing that out. Done.

**Comment by makoto** [comment]
> > Now that you ported depending library to own files, do you still need to commit changes on `package-lock.json` ?
> 
> Thanks for pointing that out. Done.

I think you misunderstood my point. I asked not to commit the change to `package-lock.json` if you haven't made any changes to `package.json` not deleting the entire file

**Comment by janus** [comment]
> > > Now that you ported depending library to own files, do you still need to commit changes on `package-lock.json` ?
> > 
> > 
> > Thanks for pointing that out. Done.
> 
> I think you misunderstood my point. I asked not to commit the change to `package-lock.json` if you haven't made any changes to `package.json` not deleting the entire file

Okay. I have returned the file. Thanks.

**Comment by makoto** [comment]
> Okay. I have returned the file. Thanks.

no, it's still wrong. Your PR should not include any diff in package-lock.json as you are not adding nor deleting any modules on package.json

**Comment by janus** [comment]
> > Okay. I have returned the file. Thanks.
> 
> no, it's still wrong. Your PR should not include any diff in package-lock.json as you are not adding nor deleting any modules on package.json

But I deleted a module from it. Groestl-hash-js was there before I copied and localized it.

**Comment by makoto** [comment]
> But I deleted a module from it. Groestl-hash-js was there before I copied and localized it.

so you removed `Groestl-hash-js` from `package.json`?
If so, how come the change is not committed?

**Comment by janus** [comment]
> > But I deleted a module from it. Groestl-hash-js was there before I copied and localized it.
> 
> so you removed `Groestl-hash-js` from `package.json`?
> If so, how come the change is not committed?

I erroneously deleted `package.json ` , I had to pull from master.  And I  have squashed,  what  should I do now to get this issue resolved?

**Comment by makoto** [comment]
You can manually copy package.json from `dev` branch I guess.

**Comment by makoto** [comment]
I am confused. `Groestl-hash-js` does not seem t exist on `master`. I am guessing you added that in your branch, then deleted. If that's the case, neither `package.json` nor `package-lock.json` should contain any changes.

**Comment by janus** [comment]
> I am confused. `Groestl-hash-js` does not seem t exist on `master`. I am guessing you added that in your branch, then deleted. If that's the case, neither `package.json` nor `package-lock.json` should contain any changes.

Yes, that is correct. Should I re-base to master? Or copy package.json and package-lock.json from master and replace my branch with them?

**Comment by makoto** [comment]
I don't really use `rebase`. I would think the latter just works

**Comment by janus** [comment]
> > > Now that you ported depending library to own files, do you still need to commit changes on `package-lock.json` ?
> > 
> > 
> > Thanks for pointing that out. Done.
> 
> I think you misunderstood my point. I asked not to commit the change to `package-lock.json` if you haven't made any changes to `package.json` not deleting the entire file

Okay. I have returned the file. Thanks.

> I don't really use `rebase`. I would think the latter just works

I followed option 2. Thanks. Please review.

**Comment by makoto** [comment]
Note to self, the file size is 346623 -> 377973, about 30k growth

**Pull request merged**